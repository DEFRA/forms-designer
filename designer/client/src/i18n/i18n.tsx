import i18next, { type InitOptions, type TOptions } from 'i18next'
import Backend from 'i18next-http-backend'
import upperFirst from 'lodash/upperFirst.js'
import * as React from 'react'

import enCommonTranslations from '~/src/i18n/translations/en.translation.json'

const interpolationFormats = {
  capitalise: (value) => upperFirst(value)
}

const DEFAULT_SETTINGS: InitOptions = {
  lng: 'en',
  fallbackLng: 'en',
  debug: false,
  interpolation: {
    escapeValue: false,
    format: function (value, format, lng) {
      return interpolationFormats[format]?.(value) ?? value
    }
  },
  resources: {
    en: {
      translation: enCommonTranslations
    }
  },
  backend: {
    loadPath: '/assets/translations/{{lng}}.{{ns}}.json'
  }
}

export const initI18n = (
  i18n: typeof i18next = i18next,
  settings = DEFAULT_SETTINGS
) => i18n.use(Backend).init(settings)

export type I18n = (text: string, options?: any) => string

export const i18n: I18n = (text: string, options: TOptions) => {
  return i18next.t(text, options)
}

export interface WithI18nProps {
  i18n: I18n
}

export const withI18n = <P extends WithI18nProps>(
  Component: React.ComponentType<P>
) => {
  return function WithI18n(props: Omit<P, keyof WithI18nProps>) {
    return <Component {...(props as P)} i18n={i18n} />
  }
}

export const withI18nRef = (WrappedComponent) => {
  function WithI18n({ forwardedRef, ...rest }) {
    return <WrappedComponent {...rest} i18n={i18n} ref={forwardedRef} />
  }

  const forwardRef = (props, ref) =>
    React.createElement(
      WithI18n,
      Object.assign({}, props, { forwardedRef: ref })
    )
  return React.forwardRef(forwardRef)
}
