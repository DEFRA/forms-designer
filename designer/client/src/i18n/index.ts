export * from '~/src/i18n/i18n.jsx'
