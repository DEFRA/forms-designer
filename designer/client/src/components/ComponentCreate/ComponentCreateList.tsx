import {
  ComponentTypes,
  controllerNameFromPath,
  ControllerType,
  hasComponents,
  hasContent,
  hasInputField,
  hasSelectionFields,
  type ComponentDef,
  type Page
} from '@defra/forms-model'
import React, { useMemo } from 'react'

import { i18n } from '~/src/i18n/i18n.jsx'

interface Props {
  page: Page
  onSelectComponent: (type: ComponentDef) => void
}

export const ComponentCreateList = (props: Readonly<Props>) => {
  const { page, onSelectComponent } = props
  const controller = controllerNameFromPath(page.controller)

  // Allow component pages to add input + selection fields
  const isComponentPage =
    controller === ControllerType.Page ||
    controller === ControllerType.FileUpload

  const componentList = useMemo(() => {
    return [...structuredClone(ComponentTypes)].sort(
      ({ type: typeA }, { type: typeB }) => typeA.localeCompare(typeB)
    )
  }, [])

  const { contentFields, selectionFields, inputFields } = useMemo(() => {
    return {
      contentFields: componentList.filter(hasContent),
      selectionFields: componentList.filter(hasSelectionFields),
      inputFields: componentList.filter(hasInputField)
    }
  }, [componentList])

  if (!hasComponents(page)) {
    return null
  }

  return (
    <div className="govuk-form-group">
      <div className="govuk-hint">{i18n('component.create_info')}</div>
      <ol className="govuk-list">
        <li>
          <h3 className="govuk-heading-s">{i18n('Content')}</h3>
          <div className="govuk-hint">
            {i18n('component.contentfields_info')}
          </div>
          <ol className="govuk-list">
            {contentFields.map((component) => (
              <li key={component.name}>
                <p className="govuk-body govuk-!-margin-bottom-2">
                  <a
                    className="govuk-link"
                    href="#0"
                    onClick={(e) => {
                      e.preventDefault()
                      onSelectComponent(component)
                    }}
                  >
                    {i18n(`fieldTypeToName.${component.type}`)}
                  </a>
                </p>
                <div className="govuk-hint govuk-!-margin-top-2">
                  {i18n(`fieldTypeToName.${component.type}_info`)}
                </div>
              </li>
            ))}
          </ol>
          {isComponentPage && (
            <hr className="govuk-section-break govuk-section-break--m govuk-section-break--visible" />
          )}
        </li>
        {isComponentPage && (
          <>
            <li>
              <h3 className="govuk-heading-s">{i18n('Input fields')}</h3>
              <div className="govuk-hint">
                {i18n('component.inputfields_info')}
              </div>
              <ol className="govuk-list">
                {inputFields.map((component) => (
                  <li key={component.type}>
                    <p className="govuk-body govuk-!-margin-bottom-2">
                      <a
                        href="#0"
                        className="govuk-link"
                        onClick={(e) => {
                          e.preventDefault()
                          onSelectComponent(component)
                        }}
                      >
                        {i18n(`fieldTypeToName.${component.type}`)}
                      </a>
                    </p>
                    <div className="govuk-hint govuk-!-margin-top-2">
                      {i18n(`fieldTypeToName.${component.type}_info`)}
                    </div>
                  </li>
                ))}
              </ol>
              <hr className="govuk-section-break govuk-section-break--m govuk-section-break--visible" />
            </li>
            <li>
              <h3 className="govuk-heading-s">{i18n('Selection fields')}</h3>
              <div className="govuk-hint">
                {i18n('component.selectfields_info')}
              </div>
              <ol className="govuk-list">
                {selectionFields.map((component) => (
                  <li key={component.type}>
                    <p className="govuk-body govuk-!-margin-bottom-2">
                      <a
                        href="#0"
                        className="govuk-link"
                        onClick={(e) => {
                          e.preventDefault()
                          onSelectComponent(component)
                        }}
                      >
                        {i18n(`fieldTypeToName.${component.type}`)}
                      </a>
                    </p>
                    <div className="govuk-hint govuk-!-margin-top-2">
                      {i18n(`fieldTypeToName.${component.type}_info`)}
                    </div>
                  </li>
                ))}
              </ol>
            </li>
          </>
        )}
      </ol>
    </div>
  )
}
