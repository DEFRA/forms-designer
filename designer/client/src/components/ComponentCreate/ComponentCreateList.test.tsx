import {
  ComponentType,
  ControllerType,
  PageTypes,
  getPageDefaults
} from '@defra/forms-model'
import { screen } from '@testing-library/dom'
import { render } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'

import { ComponentCreateList } from '~/src/components/ComponentCreate/ComponentCreateList.jsx'
import { i18n } from '~/src/i18n/i18n.jsx'

describe('Component create list: Supported page types', () => {
  const supported = [
    ControllerType.Content,
    ControllerType.Start,
    ControllerType.Page,
    ControllerType.FileUpload,
    ControllerType.Repeat
  ]

  it.each(
    PageTypes.filter(({ controller }) => {
      return !!controller && supported.includes(controller)
    })
  )("should render supported page type '$controller'", (page) => {
    const { container } = render(
      <ComponentCreateList page={page} onSelectComponent={jest.fn()} />
    )

    const $list = container.querySelector('ol')
    expect($list).toBeInTheDocument()
  })

  it.each(
    PageTypes.filter(({ controller }) => {
      return !!controller && !supported.includes(controller)
    })
  )("should not render unsupported page type '$controller'", (page) => {
    const { container } = render(
      <ComponentCreateList page={page} onSelectComponent={jest.fn()} />
    )

    const $list = container.querySelector('ol')
    expect($list).not.toBeInTheDocument()
  })
})

describe.each([
  [
    ControllerType.Content,
    {
      content: [
        ComponentType.Details,
        ComponentType.Html,
        ComponentType.InsetText,
        ComponentType.List
      ],
      selection: undefined,
      input: undefined
    }
  ],
  [
    ControllerType.Start,
    {
      content: [
        ComponentType.Details,
        ComponentType.Html,
        ComponentType.InsetText,
        ComponentType.List
      ],
      selection: undefined,
      input: undefined
    }
  ],
  [
    ControllerType.Page,
    {
      content: [
        ComponentType.Details,
        ComponentType.Html,
        ComponentType.InsetText,
        ComponentType.List
      ],
      selection: [
        ComponentType.AutocompleteField,
        ComponentType.CheckboxesField,
        ComponentType.RadiosField,
        ComponentType.SelectField,
        ComponentType.YesNoField
      ],
      input: [
        ComponentType.DatePartsField,
        ComponentType.EmailAddressField,
        ComponentType.MonthYearField,
        ComponentType.MultilineTextField,
        ComponentType.NumberField,
        ComponentType.TelephoneNumberField,
        ComponentType.TextField,
        ComponentType.UkAddressField
      ]
    }
  ],
  [
    ControllerType.FileUpload,
    {
      content: [
        ComponentType.Details,
        ComponentType.Html,
        ComponentType.InsetText,
        ComponentType.List
      ],
      selection: [],
      input: [ComponentType.FileUploadField]
    }
  ],
  [ControllerType.Summary, null],
  [ControllerType.Status, null]
])('Component create list: %s', (controller, options) => {
  const onSelectComponent = jest.fn()

  beforeEach(() => {
    const page = getPageDefaults({ controller })

    render(
      <ComponentCreateList page={page} onSelectComponent={onSelectComponent} />
    )
  })

  describe.each([
    {
      name: 'Content',
      components: options?.content
    },
    {
      name: 'Selection',
      components: options?.selection
    },
    {
      name: 'Input',
      components: options?.input
    }
  ])('$name', ({ name, components }) => {
    if (components?.length) {
      it(`should render '${name}' components`, () => {
        const $heading = screen.getByRole('heading', {
          name: 'Content'
        })

        const $list = $heading.parentElement?.querySelector('ol')
        expect($list).toBeInTheDocument()
      })

      describe.each(components)('%s component', (type) => {
        it('should render link to create component', () => {
          const $heading = screen.getByRole('heading', { name })

          const $list = $heading.parentElement?.querySelector('ol')
          const $link = screen.getByRole('link', {
            name: i18n(`fieldTypeToName.${type}`)
          })

          expect($list).toContainElement($link)
        })

        it('should run onSelectComponent callback on link click', async () => {
          const $link = screen.getByRole('link', {
            name: i18n(`fieldTypeToName.${type}`)
          })

          await userEvent.click($link)
          expect(onSelectComponent).toHaveBeenCalledWith(
            expect.objectContaining({
              type
            })
          )
        })
      })
    } else if (components) {
      it(`should render '${name}' components unavailable`, () => {
        const $heading = screen.getByRole('heading', { name })

        const $list = $heading.parentElement?.querySelector('ol')
        const $hint = $heading.parentElement?.querySelector('.govuk-hint')

        expect($list).not.toBeInTheDocument()
        expect($hint).toHaveTextContent(
          `No ${name.toLowerCase()} components available`
        )
      })
    } else {
      it(`should not render '${name}' components`, () => {
        expect(screen.queryByRole('heading', { name })).not.toBeInTheDocument()
      })
    }
  })
})
