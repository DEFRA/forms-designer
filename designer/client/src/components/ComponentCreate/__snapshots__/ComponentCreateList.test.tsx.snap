// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentCreateList should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group component-create__list"
    data-testid="component-create-list"
  >
    <h1
      class="govuk-hint"
    >
      Select a component to add to your page. If your page has only one component, the component title does not show on the page. To name or describe the component, use the page title instead.
    </h1>
    <ol
      class="govuk-list"
    >
      <li
        class="component-create__list__item"
      >
        <h2
          class="govuk-heading-s"
        >
          Content
        </h2>
        <div
          class="govuk-hint"
        >
          Content components allow you to add information to your page
        </div>
        <ol
          class="govuk-list"
        >
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Details
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              An expandable title that lets users reveal more detailed information only if they need it
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Guidance
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A block of text that you can apply basic HTML to, such as text formatting and hyperlinks
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Inset text
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A block of text that differentiates it from surrounding content, for example for quotes
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                List
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A series of bullet points
            </div>
          </li>
        </ol>
        <hr
          class="govuk-section-break govuk-section-break--m govuk-section-break--visible"
        />
      </li>
      <li
        class="component-create__list__item"
      >
        <h2
          class="govuk-heading-s"
        >
          Input fields
        </h2>
        <div
          class="govuk-hint"
        >
          Input fields allow you to ask users to enter information
        </div>
        <ol
          class="govuk-list"
        >
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Date
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              Allows users to manually enter a day, month and year
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Email address
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A text box that users can enter an email address into
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Month & year
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              Allows users to manually enter a month and year
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Textarea
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A text box for users to enter multiple lines of text into, for example as 'Additional info'
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Number
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A text box that users can only enter numbers into
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Telephone number
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A text box that users can enter numbers and some symbols into, such as '+'
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Text input
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A text box that users can enter letters, numbers and symbols into
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                UK address
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              Allows users to enter their street address, town or city and postcode
            </div>
          </li>
        </ol>
        <hr
          class="govuk-section-break govuk-section-break--m govuk-section-break--visible"
        />
      </li>
      <li
        class="component-create__list__item"
      >
        <h2
          class="govuk-heading-s"
        >
          Selection fields
        </h2>
        <div
          class="govuk-hint"
        >
          Selection fields allow you to provide lists of answers for users to choose from
        </div>
        <ol
          class="govuk-list"
        >
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Autocomplete
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              Sets a list of values to suggest to users as they complete this field
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Checkboxes
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A list for users to select multiple answers
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Radios
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A list for users to select one answer
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                Select
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              A drop-down list for users to select one answer
            </div>
          </li>
          <li>
            <p
              class="govuk-body govuk-!-margin-bottom-2"
            >
              <a
                class="govuk-link"
                href="#0"
              >
                YesNo
              </a>
            </p>
            <div
              class="govuk-hint govuk-!-margin-top-2"
            >
              Allows users to answer either 'Yes' or 'No'
            </div>
          </li>
        </ol>
      </li>
    </ol>
  </div>
</DocumentFragment>
`;
